### ISSUE TITLE
Intermittent test failure in `checksum_sanity_32`

### ISSUE THREAD

**Issue by rocallahan** [rr-debugger/rr#3589]
I saw this locally. My trace contains
```
{
  real_time:40423.732678 global_time:458, event:`SYSCALL: openat' (state:ENTERING_SYSCALL) tid:313736, ticks:46419
eax:0xffffffda ecx:0x804a2f0 edx:0xc2 ebx:0xffffff9c esp:0x681ffe10 ebp:0x0 esi:0x180 edi:0x0 eip:0x7000000e eflags:0x246 xcs:0x23 xss:0x2b xds:0x2b
 xes:0x2b xfs:0x0 xgs:0x63 orig_eax:0x127
}
{
  real_time:40423.746741 global_time:459, event:`SYSCALL: openat' (state:EXITING_SYSCALL) tid:313736, ticks:46419
eax:0x7 ecx:0x804a2f0 edx:0xc2 ebx:0xffffff9c esp:0x681ffe10 ebp:0x0 esi:0x180 edi:0x0 eip:0x7000000e eflags:0x246 xcs:0x23 xss:0x2b xds:0x2b xes:0x
2b xfs:0x0 xgs:0x63 orig_eax:0x127
  { tid:313736, addr:0x7000407e, length:0x8 }
  { tid:313736, addr:0x7000408e, length:(nil) }
}
{
  real_time:40423.746763 global_time:460, event:`SYSCALLBUF_ABORT_COMMIT' tid:313736, ticks:46419
}
```
Normally there are no `SYSCALLBUF_ABORT_COMMIT`s in the trace. I guess the tracee just randomly got descheduled during the may-block `openat()` call, which normally doesn't block at all.

I should be able to write a testcase that reliably blocks in `openat`, which hopefully will reproduce the problem reliably.

**Comment by rocallahan** [comment]
One problem is that during recording, `save_interrupted_syscall_ret_in_syscallbuf()` writes `syscallbuf_record::ret` as a memory effect of the syscall-exit for the interrupted syscall, but the actual write to that memory doesn't occur until the syscallbuf code does it in `untraced_syscall_full()` as we execute code in the next trace event. So during replay at the end of the syscall-exit event that write has happened, but during recording it hasn't.

**Comment by rocallahan** [comment]
Another problem is that during recording we set `syscallbuf_hdr::notify_on_syscall_hook_exit` to 1 when we write the ABORT_COMMIT record in `RecordSession::desched_state_changed()`. This write is not recorded though; during replay we set that flag when we start running towards a `rrcall_notify_syscall_hook_exit` syscall-entry. In code like the buffering code for `openat`, that can be quite a bit later because after a successful open that buffering code does some more system calls that can be traced. For the events in between the memory checksums won't match.
